### ISSUE TITLE
EOF issue

### ISSUE THREAD

**Issue by fse-it** [jborean93/smbprotocol#20]
Hi Jordan, great work on this repo esp the new 'high level" addition.

I attempted to use open_file with v1.0.0 and got error like

       smbprotocol.exceptions.SMBResponseException: Received unexpected status from the server: (3221225489) STATUS_END_OF_FILE: 0xc0000011

which from [this site](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-cifs/8f11e0f3-d545-46cc-97e6-f00569e3e1bc)  means "Attempted to read beyond the end of the file."

The code almost identical to example...  python3 code;

import smbclient
smbclient.register_session("server", username="user@domain", password="pass")
with smbclient.open_file(file_path) as fd:
      print (fd.read())

I have verified the connection is fine as stat() works fine on file_path which is a .txt file.

kind regards,
AK

**Comment by jborean93** [comment]
That's interesting, are there any details on the file that you can share? Does it have any data or is it a certain file size, just printing `smbclient.stat(file_path)` and sharing that info should help. Also what is the SMB Server you are targeting, is it just a Windows host or some other implementation?

According to [MS-SMB2](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/21e8b343-34e9-4fca-8d93-03dd2d3e961e) a server should only return `STATUS_END_OF_FILE` if `MinimumCount` is set and the number of bytes is less than the minimum set.

> If the read fails, the server MUST fail the request using the error code received from the read operation. If the read returns fewer bytes than specified by the MinimumCount field of the request, the server MUST fail the request with STATUS_END_OF_FILE.

The read operation that `smbclient` [calls](https://github.com/jborean93/smbprotocol/blob/b6abd4ae092ee01a3ab2a6b8999237b293151b90/smbclient/_io.py#L417) does not set the `MinimumCount` field so theorectically this shouldn't fail on a Windows SMB server but let's try and investigate further.

**Comment by fse-it** [comment]
Thanks for reply.

The device is not a windows server but a Dell EMC Isilon (not too sure but I think its OneFS which is based on FreeBSD so guessing closer to Samba).

The stat for file output is, it is a csv file of some size;

       SMBStatResult(st_mode=33206, st_ino=5623801167, st_dev=10131971, st_nlink=1, st_uid=0, st_gid=0, st_size=32190726, st_atime=1575367282.0, st_mtime=1575367282.0, st_ctime=1575367941.070796, st_chgtime=1575367943.8862367, st_atime_ns=1575367282000000000, st_mtime_ns=1575367282000000000, st_ctime_ns=1575367941070795900, st_chgtime_ns=1575367943886236600, st_file_attributes=32, st_reparse_tag=0)

**Comment by jborean93** [comment]
So my guess is that the Dell SMB implementation is not 100% up to spec and if the buffer size exceeds the end of file then it will fail with this error. It's definitely something we can handle but I would like to figure out what exactly causes the error.

It might be easier to replicate on a smaller file. Are you able to try out the following:

```
import smbclient

smbclient.register_session("server", username="user@domain", password="pass")

# This should be a path where you can create this file
file_path = r"\\server\share\test.txt"

with smbclient.open_file(file_path, mode='w') as fd:
    fd.write(u"Test")

with smbclient.open_file(file_path, mode='rb', buffering=0) as fd:
    raw_fd = fd.fd

    # I expect this to be 4
    print("EOF: %s" % raw_fd.end_of_file)

    # I expect this to pass and print b"Tes"
    print(raw_fd.read(0, 3))

    # I also expect this to pass and print b"Test"
    print(raw_fd.read(0, 4))

    # This should work on Windows but I expect the Dell implementation to fail with EOF
    print(raw_fd.read(0, 5))
```

If the read from 0-5 fails we can probably add a try/except `STATUS_END_OF_FILE` to the read call to cover implementations like this but let me know how you go.

**Comment by fse-it** [comment]
That would be great to find a fix which may help other "shortcut" implementations out there...

Output of program is

```
EOF: 4
b'Tes'
b'Test'
b'Test'
Traceback (most recent call last):
  File "t.py", line 30, in <module>
    print(raw_fd.read(0, 5, min_length=5))
  File "/usr/local/lib/python3.6/dist-packages/smbprotocol/open.py", line 1305, in read
    return self._read_response(request, wait)
  File "/usr/local/lib/python3.6/dist-packages/smbprotocol/open.py", line 1311, in _read_response
    response = self.connection.receive(request, wait=wait)
  File "/usr/local/lib/python3.6/dist-packages/smbprotocol/connection.py", line 763, in wrapped
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/smbprotocol/connection.py", line 1075, in receive
    raise SMBResponseException(response, status)
smbprotocol.exceptions.SMBResponseException: Received unexpected status from the server: (3221225489) STATUS_END_OF_FILE: 0xc0000011

```

**Comment by jborean93** [comment]
Actually that isn't what I expected unfortunately, I expected the 3rd read to fail with the EOF but the server is doing what Windows would do.

Maybe this is only a problem with files or a certain size. We can prove that by running

```
import smbclient

smbclient.register_session("server", username="user@domain", password="pass")

file_path = r"\\server\share\test.txt"

with smbclient.open_file(file_path, mode='w') as fd:
    fd.write(u"Test")

with smbclient.open_file(file_path, mode='r') as fd:
    print(fd.read())
```

According to your original bug report it was just a standard `.read()` on an opened file that caused the EOF. If it doesn't fail here then something else about the original issue is at play. If it does fail then I honestly have no idea but at least we can rule out something.

**Comment by fse-it** [comment]
Unfortunately, the latest code succeeds with return "Test" no errors :(

Could it be permissions? One difference between by original file and the "test.txt" file is that the original is on a readonly share therefore I had to test on a RW folder on same server instead?

**Comment by jborean93** [comment]
It's worth a try, the only reason why I had the write part is so we can test with a file off a known size. Trying out anything you can and finding the differences will definitely help track it down. What I would try out

* Try to place that 4 byte file we test and try out the tests again
* Try to read another file on the same share
* Try to read the same file on another share that is RW
* Try to read the same file on another server

Finally another thing you can do is to either capture the traffic over port 445 and share that. You could also enable debug logging in smbprotocol and share that info. If you wanted to rely on the smbprotocol logging just add the following to the top of your script

```
import logging

logging.basicConfig(level=logging.DEBUG,
                    filename='smb.log',
                    format='%(asctime)s %(message)s')
```

The file `smb.log` will contain the SMB messages sent to and from the server so be careful that you remove any sensitive information before sharing the file.

**Comment by fse-it** [comment]
Hi Jordan,

Some more investigation reveals that the file size is the main factor... at 

131,07**2** bytes the file transfers fine however just adding more char to tip it over 131,07**3** bytes returns the STATUS_END_OF_FILE error.


broken stat() output
`SMBStatResult(st_mode=33206, st_ino=5632057213, st_dev=10131971, st_nlink=1, st_uid=0, st_gid=0, st_size=131073, st_atime=1575529024.0, st_mtime=1575529024.0, st_ctime=1575367941.0, st_chgtime=1575529026.6656456, st_atime_ns=1575529024000000000, st_mtime_ns=1575529024000000000, st_ctime_ns=1575367941000000000, st_chgtime_ns=1575529026665645700, st_file_attributes=32, st_reparse_tag=0)`

Ok size stat() output
`SMBStatResult(st_mode=33206, st_ino=5642572094, st_dev=10131971, st_nlink=1, st_uid=0, st_gid=0, st_size=131072, st_atime=1575529117.0, st_mtime=1575529117.0, st_ctime=1575367941.0, st_chgtime=1575529119.746629, st_atime_ns=1575529117000000000, st_mtime_ns=1575529117000000000, st_ctime_ns=1575367941000000000, st_chgtime_ns=1575529119746629000, st_file_attributes=32, st_reparse_tag=0)`


Are you able to reproduce?

**Comment by jborean93** [comment]
Bingo! I am coming across the same error now with a file of that size. Will investigate it further and figure out why it is happening and how we can fix it.

**Comment by jborean93** [comment]
Found the issue, each read request is done in `65536` byte increments (by default) and `131072` just fits under 2 calls. Due to a bug that was incrementing the offset based on the total bytes read during a `read()` operation and not the bytes read for each SMB2 read response, the 3rd call would have an offset that is too large for the file.

The PR https://github.com/jborean93/smbprotocol/pull/22 should fix this up for you, if you do get a chance it would be great if you could test it out when you can.

**Comment by jborean93** [comment]
Thanks for sticking with this and figuring out what caused the issue. I've merged #22 and it will be available when I create the 1.0.1 release.

**Comment by fse-it** [comment]
Thank you for quick fix. I can confirm fix. Keep up good work mate.
